crates/iceberg/src/catalog.rs

```diff
@@ -0,0 +1,149 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Catalog API for Apache Iceberg
+
+use crate::spec::{PartitionSpec, Schema, SortOrder};
+use crate::table::Table;
+use crate::Result;
+use async_trait::async_trait;
+use std::collections::HashMap;
+
+/// The catalog API for Iceberg Rust.
+#[async_trait]
+pub trait Catalog {
+    /// List namespaces from table.
+    async fn list_namespaces(&self, parent: Option<&NamespaceIdent>)
```

Comment:
From my experience in implementing icelake, we should replace `&self` with `Arc<Self>`

Reply:
Can you explain a bit more about this?

Reply:
This is inspired by the design in python implementation. We need to store a `Arc<dyn Cataloyg>` in `Table` for several things such as referesh , transaction api: https://github.com/apache/iceberg/blob/751d15337f436c9416b828e50c5373de093b32a9/python/pyiceberg/table/__init__.py#L416

This way we can only pass `Arc<Self>` in `Catalog`

Reply:
I still hesitate about this design. How about improving this in the future? I expect user to use `Arc<dyn Cataloyg>` by user and we `impl<T: Catalog + ?Sized> Catalog for Arc<T>`.

Reply:
Cool, we can implement in future.